Add support for Automatic Migrations

Abizern opened the pull request.

Abizern committed:
Add support for Automatic Migrations

Abizern committed:
Add support for Automatic Migrations

Abizern committed:
Bring the sample project into the repository

Abizern committed:
Tidy the model and remove dependency on mogenerator

It's a small project, don't want to put anybody off with unneeded dependencies

Abizern committed:
Add test targets for ARC and Non-ARC projects

Normally, I would just use one common test class, but I'll leave them as
individual ones for now to keep the tests more flexible.

Abizern committed:
Create test files that will actually be used

I prefer single test files, a public interface for a Unit Test doesn't make
sense.

Just a stub with an indication of the next test I will write.

Abizern committed:
Create a new version of the test apps model and write a failing test for migrations

Abizern committed:
Update the Test Apps to use version2 of the model

Abizern committed:
Add a Readme to the repository

Abizern committed:
Add a Readme to the repository

Abizern commented:
This has been tested, but I don't want to create a pull request until you create a branch called `tests`.

mikeabdullah commented on BSManagedDocument.m:
```diff
@@ -136,13 +136,28 @@ - (BOOL)configurePersistentStoreCoordinatorForURL:(NSURL *)storeURL
+
+    if ([self enableAutomaticMigration]) {
+        [mutableOptions setObject:@YES forKey:NSMigratePersistentStoresAutomaticallyOption];
+        [mutableOptions setObject:@YES forKey:NSInferMappingModelAutomaticallyOption];
```
One thing that bothers me is this currently overrides such options if a subclass were to pass them in

Abizern replied:
In which case override `enableAutomaticMigration` to return `NO` in the subclass and pass in whatever you want. The idea is to create a sensible default.

mikeabdullah commented:
> I don't want to create a pull request until you create a branch called tests.

Voila, one tests branch.

mikeabdullah committed:
Only supply automatic migration options when configuring persistent store for reading
Makes no sense to migrate a store that's being created, right? I also like that this neatly solves the issue of overriding those settings should a subclass try to do so from *its* override of -configurePersistentStore…

mikeabdullah committed:
Rename -enableAutomaticMigration as +requestsAutomaticMigration

mikeabdullah commented:
I'm still struggling to decide if I actually want to have this baked in as a feature. In the meantime, have made some changes on my `automatic-migration` branch which hopefully you're happy to pull back into yours

Abizern committed:
Add a .gitignore file

I'm so used to my system wide one, I forget to add them
to repositories, sometimes.

Abizern commented:
Nice, thanks.

mikeabdullah commented:
So I guess my big issue is that this is only really useful during initial development of an app. Once you start shipping to real customers, it seems rather rude to silently upgrade documents upon simply opening them. Instead you want some sort of UI prompt, or to defer the migration until first save, I think.

Come to think of it, I wonder what happens when combined with Versions. Documents opened from browsing through Versions are read-only so migration would be impossible. I wonder if Core Data ignores the request though, or fails with an error.

It seems to me it would be better just to whack a snippet of code into the README that devs can quickly paste into their `BSManagedDocument`-subclass in order to switch on automatic migration. How about you?

Abizern commented:
I see your point.

You have more experience that I about creating general use frameworks - so I see your point. My current use case is because I'm in development, I I just want to be able to migrate my documents without worrying about them. So currently this works for me.

That said - I think there should be some built in support for migrations. Maybe even a delegate protocol that can be queried when a the model changes.

mikeabdullah commented:
> Maybe even a delegate protocol that can be queried when a the model changes.

I'm confused. You'd set up some custom object as the document's delegate? That seems freaky. Or something else?

Abizern commented:
I was thinking out loud. 

I have an idea. If it works I'll push it for comment. 

## 

Abizer

On 28 Jan 2013, at 17:43, Mike Abdullah notifications@github.com wrote:

> Maybe even a delegate protocol that can be queried when a the model changes.
> 
> I'm confused. You'd set up some custom object as the document's delegate? That seems freaky. Or something else?
> 
> —
> Reply to this email directly or view it on GitHub.

Abizern committed:
revert out automatic migration changes for now

Abizern committed:
Revert automatic migration code

Abizern committed:
synchronize with tests

Abizern committed:
Remove scaffolding

Abizern committed:
Pull in upstream changes
